stdlib/public/core/Span/Span.swift

```diff
@@ -0,0 +1,832 @@
+  ) throws(E) -> Result {
+    guard let pointer = _pointer, count > 0 else {
+      return try body(.init(start: nil, count: 0))
+    }
+    let binding = Builtin.bindMemory(pointer._rawValue, count._builtinWordValue, Element.self)
+    defer { Builtin.rebindMemory(pointer._rawValue, binding) }
+    return try body(.init(start: .init(pointer._rawValue), count: count))
+  }
+}
+
+extension Span where Element: BitwiseCopyable {
+
+  //FIXME: mark closure parameter as non-escaping
+  /// Calls the given closure with a pointer to the underlying bytes of
+  /// the viewed contiguous storage.
+  ///
+  /// The buffer pointer passed as an argument to `body` is valid only
+  /// during the execution of `withUnsafeBytes(_:)`.
+  /// Do not store or return the pointer for later use.
+  ///
+  /// - Parameter body: A closure with an `UnsafeRawBufferPointer`
+  ///   parameter that points to the viewed contiguous storage.
+  ///   If `body` has a return value, that value is also
+  ///   used as the return value for the `withUnsafeBytes(_:)` method.
+  ///   The closure's parameter is valid only for the duration of
+  ///   its execution.
+  /// - Returns: The return value of the `body` closure parameter.
+  @_alwaysEmitIntoClient
+  public func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
+  ) throws(E) -> Result {
+    try RawSpan(_unsafeSpan: self).withUnsafeBytes(body)
+  }
+}
+
+// `first` and `last` can't exist where Element: ~Copyable
+// because we can't construct an Optional of a borrow
+extension Span where Element: Copyable {
+
+  /// The first element in the span.
+  ///
+  /// If the span is empty, the value of this property is `nil`.
+  ///
+  /// - Returns: The first element in the span, or `nil` if empty
+  @_alwaysEmitIntoClient
+  public var first: Element? {
+    isEmpty ? nil : self[unchecked: 0]
+  }
+
+  /// The last element in the span.
+  ///
+  /// If the span is empty, the value of this property is `nil`.
+  ///
+  /// - Returns: The last element in the span, or `nil` if empty
+  @_alwaysEmitIntoClient
+  public var last: Element? {
+    isEmpty ? nil : self[unchecked: count &- 1]
+  }
+}
+
+extension Span where Element: ~Copyable {
+  /// Returns a Boolean value indicating whether two `Span` instances
+  /// refer to the same region in memory.
+  @_alwaysEmitIntoClient
+  public func isIdentical(to other: Self) -> Bool {
+    (self._pointer == other._pointer) && (self._count == other._count)
+  }
+
+  /// Returns true if the memory represented by `span` is a subrange of
+  /// the memory represented by `self`
+  ///
+  /// Parameters:
+  /// - span: a span of the same type as `self`
+  /// Returns: whether `span` is a subrange of `self`
+  @_alwaysEmitIntoClient
+  public func isWithin(_ span: borrowing Self) -> Bool {
+    if _count > span._count { return false }
+    if _count == 0 { return true }
```

Comment:
This seems surprising – why is a zero-length span always considered a subrange of every other span?

Reply:
We could make the effort of making sure that the span's pointer is within the range of addresses of the other span, but I'm not convinced it's worth the computational complexity. Otherwise, the logic is that the empty set is included in any other set.

Reply:
After trying both ways, I changed the implementation to be stricter, as in a zero-length span is only a subrange of another if the addresses actually match (or they're both `nil`.) The additional complexity was actually minimal.